### PR TITLE
Revamped Laphine UIs

### PR DESCRIPTION
**Pull request by Lemongrass3110**
* **Addressed Issue(s)**: #3302 and #4348 

* **Server Mode**: Renewal

* **Description of Pull Request**: 
Thanks for the initial release by @Cydh in #4348 and everyone that contributed to it.

All existing data was migrated and cleaned up where necessary.
Thanks to @Everade for his help here.

Laphine UIs are now fully yamlified and not dependent on the script engine.
They make use of new item group features and of the already existing random option group feature.
This way they will be far easier to be maintained, even though they are a little less customize able.

### CODE REVIEW

**Comment by Pokye** [comment]
Thanks sir Lemongrass <3

**Comment by SeyrenDev** [comment]
@Lemongrass3110  serious question, for those who already have the lapine system of cydh, do we have to remove the previous git from cyndh or just update the gith ?

**Comment by Lemongrass3110** [comment]
> @Lemongrass3110 serious question, for those who already have the lapine system of cydh, do we have to remove the previous git from cyndh or just update the gith ?

I cannot give you a 100% answer. This depends very hard on your setup (how you set up your private source control).
Basically for most users it will be undoing cydh's version and updating rAthena after it was merged to master.

**Comment by zdurexz** [comment]
Thanks sir Lemongrass , waiting for this

**Comment by Lemongrass3110** [comment]
Sorry that it took a little longer than expected everyone!

**Comment by SeyrenDev** [comment]
> > @Lemongrass3110 pregunta seria, para aquellos que ya tienen el sistema lapine de cydh, ¿tenemos que eliminar el git anterior de cyndh o simplemente actualizar el gith?
> 
> No puedo darte una respuesta al 100%. Esto depende mucho de su configuración (cómo configura su control de código fuente privado). Básicamente para la mayoría de los usuarios, será deshacer la versión de cydh y actualizar rAthena después de que se fusionó con master.

thanks for replying <3 , ok then, I have to start from 0 XD!

**Comment by JohnnyPlayy** [comment]
What a beautiful job! 😻

**Comment by laziem** [comment]
Thank you for the release sir Lemongrass.

Got two question, first about script support that previously on cydh initial release, is it possible it be added in this release? So we don't have to edit src for every item group, can simply use `getitem itemid`. And we can also set different RandomOptionGroup for different equipment from single laphine item, adding specialeffect, etc.

Another question, for itemid: 23545 (Silver_Statue). This item supposed to give 2 RandomOption to Cursed Knight Shield, and during the process, there's a small chance to transform the shield into Bloody Knight Shield or Purified Knight Shield. The effect currently getting random one of the three shield only.

**Comment by dimasshotta** [comment]
image
I found a bug, when using Shadow_Refine_Hammer the refine level only grant +1 dan when using Shadow_Random_Mix the refine level exceed the maximum level.
supposedly the refine level is random beetween 7 - 10

**Comment by Lemongrass3110** [comment]
> Thank you for the release sir Lemongrass.
> 
> Got two question, first about script support that previously on cydh initial release, is it possible it be added in this release? So we don't have to edit src for every item group, can simply use `getitem itemid`. And we can also set different RandomOptionGroup for different equipment from single laphine item, adding specialeffect, etc.
> 
> Another question, for itemid: 23545 (Silver_Statue). This item supposed to give 2 RandomOption to Cursed Knight Shield, and during the process, there's a small chance to transform the shield into Bloody Knight Shield or Purified Knight Shield. The effect currently getting random one of the three shield only.

We can add script support later on, but I do not want to add it like it was in the old version, especially for Laphine Upgrade.
You will get the inventory index of the upgraded item only. This would require #6553 merged first.

As for the editing source part I think you refer to the item group IDs. I also thought this is annoying and will look into this to not make these a source edit anymore.

What you said about different equipment from the same item this is already possible with item groups. No need for a script here, just set up the group accordingly.

And what you asked about Silver Statue is already covered as well by the item group feature:
```yaml
  - Group: SILVER_STATUE
    SubGroups:
      - SubGroup: 1
        List:
          - Item: Bloody_Knight_Shield
            Rate: 900
          - Item: Bloody_Knight_Shield_
            Rate: 99
          - Item: Bloody_Knight_Shield__
            Rate: 1
```

**Comment by dimasshotta** [comment]
> image I found a bug, when using Shadow_Refine_Hammer the refine level only grant +1 dan when using Shadow_Random_Mix the refine level exceed the maximum level. supposedly the refine level is random beetween 7 - 10
it's fixed already, i tried to change this part

	}else if( upgrade->resultRefineMaximum > 0 ){
		// If a minimum is specified it can also downgrade
		if( upgrade->resultRefineMinimum ){
			item->refine = rnd_value( upgrade->resultRefineMinimum, upgrade->resultRefine );
		}else{
			// Otherwise it can only be upgraded until the maximum, but not downgraded
			item->refine = rnd_value( item->refine, upgrade->resultRefine );
		}
to this:
	}else if( upgrade->resultRefineMaximum > 0 ){
		// If a minimum is specified it can also downgrade
		if( upgrade->resultRefineMinimum ){
			item->refine = rnd_value( upgrade->resultRefineMinimum, upgrade->resultRefineMaximum );
		}else{
			// Otherwise it can only be upgraded until the maximum, but not downgraded
			item->refine = rnd_value( item->refine, upgrade->resultRefineMaximum );
		}

**Comment by Pokye** [comment]
do you intend to also add **@reloadsynthesisdb** and **@reloadupgradedb** like Cydh?

**Comment by Lemongrass3110** [comment]
> do you intend to also add **@reloadsynthesisdb** and **@reloadupgradedb** like Cydh? and the function npc https://github.com/rathena/rathena/blob/4eaa33c22158395c957c7ad8ab60d474a0293cdd/npc/re/other/LapineEnchants.txt Won't you need it too?

Just use `reloaditemdb` it already works.
No the function NPC is not needed.

File: db/import-tmpl/laphine_synthesis.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#
+###########################################################################
+# Laphine Synthesis Database
+###########################################################################
+#
+# Laphine Synthesis Settings
+#
+###########################################################################
+# - Item                       Item that triggers Laphine Synthesis.
+#   RewardGroup                Item Group that will given out as a reward.
```
Comment (aleos89) [inline db/import-tmpl/laphine_synthesis.yml:26]:
```suggestion
#   RewardGroup                Item Group that will be given out as a reward.
```

File: db/import-tmpl/laphine_synthesis.yml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#   RequiredRequirementsCount  How many requirements have to be fulfilled? (Default: 1)
```
Comment (aleos89) [inline db/import-tmpl/laphine_synthesis.yml:27]:
```suggestion
#   RequiredRequirementsCount  Amount of requirements that have to be fulfilled. (Default: 1)
```

File: db/import-tmpl/laphine_synthesis.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+#   MinimumRefine              Minimum refine level of the required items. (Default: 0)
+#   MaximumRefine              Maximum refine level of the required items. (Default: MAX_REFINE)
+#   Requirements:              List of possible item(s) as requirements
```
Comment (aleos89) [inline db/import-tmpl/laphine_synthesis.yml:30]:
```suggestion
#   Requirements:              List of requirement items.
```

File: db/import-tmpl/laphine_synthesis.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+#     - Item                   Item name of the required item
+#       Amount                 How many of the specific item are required? (Default: 1)
```
Comment (aleos89) [inline db/import-tmpl/laphine_synthesis.yml:32]:
```suggestion
#       Amount                 Amount if specific required item. (Default: 1)
```

File: src/map/itemdb.cpp [hunk 5]
```diff
@@ -1259,6 +1259,392 @@ int16 ItemGroupDatabase::item_exists_pc(map_session_data *sd, uint16 group_id)
...
 }
 
+const std::string LaphineSynthesisDatabase::getDefaultLocation(){
+	return std::string( db_path ) + "/laphine_synthesis.yml";
+}
+
+uint64 LaphineSynthesisDatabase::parseBodyNode( const YAML::Node& node ){
+	t_itemid item_id;
+
+	{
```
Comment (aleos89) [inline src/map/itemdb.cpp:1269]: Is there a purpose this was wrapped?

File: src/map/itemdb.cpp [hunk 6]
```diff
@@ -1259,6 +1259,392 @@ int16 ItemGroupDatabase::item_exists_pc(map_session_data *sd, uint16 group_id)
...
+		std::string name;
+
+		if( !this->asString( node, "Item", name ) ){
+			return 0;
+		}
+
+		std::shared_ptr<item_data> id = item_db.search_aegisname( name.c_str() );
+
+		if( id == nullptr ){
+			this->invalidWarning( node["Item"], "laphine_synthesis_parseBodyNode: Unknown item \"%s\".\n", name.c_str() );
```
Comment (aleos89) [inline src/map/itemdb.cpp:1279]: Were we starting to put the function name in the warnings now?

File: src/map/itemdb.cpp [hunk 7]
```diff
@@ -1259,6 +1259,392 @@ int16 ItemGroupDatabase::item_exists_pc(map_session_data *sd, uint16 group_id)
...
+LaphineSynthesisDatabase laphine_synthesis_db;
+
+const std::string LaphineUpgradeDatabase::getDefaultLocation(){
+	return std::string( db_path ) + "/laphine_upgrade.yml";
+}
+
+uint64 LaphineUpgradeDatabase::parseBodyNode( const YAML::Node& node ){
+	t_itemid item_id;
+
+	{
```
Comment (aleos89) [inline src/map/itemdb.cpp:1440]: Wrap here too. Not sure if you left for visual difference or what.

File: src/map/itemdb.cpp [hunk 8]
```diff
@@ -1259,6 +1259,392 @@ int16 ItemGroupDatabase::item_exists_pc(map_session_data *sd, uint16 group_id)
...
+
+		int64 constant;
+
+		if( !script_get_constant( ( "IG_" + name ).c_str(), &constant ) ){
+			this->invalidWarning( node["RewardGroup"], "laphine_synthesis_parseBodyNode: Unknown reward group \"%s\".\n", name.c_str() );
+			return 0;
+		}
+
+		if( !itemdb_group.exists( (uint16)constant ) ){
+			this->invalidWarning( node["RewardGroup"], "laphine_synthesis_parseBodyNode: Unknown reward group id \"%" PRId64 "\".\n", constant );
```
Comment (aleos89) [inline src/map/itemdb.cpp:1313]:
```suggestion
			this->invalidWarning( node["RewardGroup"], "laphine_synthesis_parseBodyNode: Unknown reward group ID \"%" PRId64 "\".\n", constant );
```

File: src/map/itemdb.cpp [hunk 9]
```diff
@@ -1259,6 +1259,392 @@ int16 ItemGroupDatabase::item_exists_pc(map_session_data *sd, uint16 group_id)
...
+
+			if( this->nodeExists( requirementNode, "Amount" ) ){
+				uint16 amount;
+
+				if( !this->asUInt16( requirementNode, "Amount", amount ) ){
+					return 0;
+				}
+
+				if( amount > MAX_AMOUNT ){
+					this->invalidWarning( requirementNode["Amount"], "laphine_synthesis_parseBodyNode: amount %hu is too high, capping to MAX_AMOUNT...\n", amount );
```
Comment (aleos89) [inline src/map/itemdb.cpp:1369]:
```suggestion
					this->invalidWarning( requirementNode["Amount"], "laphine_synthesis_parseBodyNode: Amount %hu is too high, capping to MAX_AMOUNT...\n", amount );
```

File: src/map/itemdb.cpp [hunk 10]
```diff
@@ -1259,6 +1259,392 @@ int16 ItemGroupDatabase::item_exists_pc(map_session_data *sd, uint16 group_id)
...
+
+	if( this->nodeExists( node, "MinimumRefine" ) ){
+		uint16 refine;
+
+		if( !this->asUInt16( node, "MinimumRefine", refine ) ){
+			return 0;
+		}
+
+		if( refine > MAX_REFINE ){
+			this->invalidWarning( node["MinimumRefine"], "laphine_synthesis_parseBodyNode: minimum refine %hu is too high, capping to MAX_REFINE...\n", refine );
```
Comment (aleos89) [inline src/map/itemdb.cpp:1394]:
```suggestion
			this->invalidWarning( node["MinimumRefine"], "laphine_synthesis_parseBodyNode: Minimum refine %hu is too high, capping to MAX_REFINE...\n", refine );
```

File: src/map/itemdb.cpp [hunk 11]
```diff
@@ -1259,6 +1259,392 @@ int16 ItemGroupDatabase::item_exists_pc(map_session_data *sd, uint16 group_id)
...
+
+	if( this->nodeExists( node, "MaximumRefine" ) ){
+		uint16 refine;
+
+		if( !this->asUInt16( node, "MaximumRefine", refine ) ){
+			return 0;
+		}
+
+		if( refine > MAX_REFINE ){
+			this->invalidWarning( node["MaximumRefine"], "laphine_synthesis_parseBodyNode: maximum refine %hu is too high, capping to MAX_REFINE...\n", refine );
```
Comment (aleos89) [inline src/map/itemdb.cpp:1413]:
```suggestion
			this->invalidWarning( node["MaximumRefine"], "laphine_synthesis_parseBodyNode: Maximum refine %hu is too high, capping to MAX_REFINE...\n", refine );
```

File: src/map/itemdb.cpp [hunk 12]
```diff
@@ -1259,6 +1259,392 @@ int16 ItemGroupDatabase::item_exists_pc(map_session_data *sd, uint16 group_id)
...
+
+	if( this->nodeExists( node, "MinimumRefine" ) ){
+		uint16 refine;
+
+		if( !this->asUInt16( node, "MinimumRefine", refine ) ){
+			return 0;
+		}
+
+		if( refine > MAX_REFINE ){
+			this->invalidWarning( node["MinimumRefine"], "laphine_upgrade_parseBodyNode: minimum refine %hu is too high, capping to MAX_REFINE...\n", refine );
```
Comment (aleos89) [inline src/map/itemdb.cpp:1519]:
```suggestion
			this->invalidWarning( node["MinimumRefine"], "laphine_upgrade_parseBodyNode: Minimum refine %hu is too high, capping to MAX_REFINE...\n", refine );
```

File: src/map/itemdb.cpp [hunk 13]
```diff
@@ -1259,6 +1259,392 @@ int16 ItemGroupDatabase::item_exists_pc(map_session_data *sd, uint16 group_id)
...
+
+	if( this->nodeExists( node, "MaximumRefine" ) ){
+		uint16 refine;
+
+		if( !this->asUInt16( node, "MaximumRefine", refine ) ){
+			return 0;
+		}
+
+		if( refine > MAX_REFINE ){
+			this->invalidWarning( node["MaximumRefine"], "laphine_upgrade_parseBodyNode: maximum refine %hu is too high, capping to MAX_REFINE...\n", refine );
```
Comment (aleos89) [inline src/map/itemdb.cpp:1538]:
```suggestion
			this->invalidWarning( node["MaximumRefine"], "laphine_upgrade_parseBodyNode: Maximum refine %hu is too high, capping to MAX_REFINE...\n", refine );
```

File: src/map/itemdb.cpp [hunk 14]
```diff
@@ -1259,6 +1259,392 @@ int16 ItemGroupDatabase::item_exists_pc(map_session_data *sd, uint16 group_id)
...
+
+	if( this->nodeExists( node, "RequiredRandomOptions" ) ){
+		uint16 amount;
+
+		if( !this->asUInt16( node, "RequiredRandomOptions", amount ) ){
+			return 0;
+		}
+
+		if( amount > MAX_ITEM_RDM_OPT ){
+			this->invalidWarning( node["RequiredRandomOptions"], "laphine_upgrade_parseBodyNode: required random option amount %hu is too high, capping to MAX_ITEM_RDM_OPT...\n", amount );
```
Comment (aleos89) [inline src/map/itemdb.cpp:1557]:
```suggestion
			this->invalidWarning( node["RequiredRandomOptions"], "laphine_upgrade_parseBodyNode: Required random option amount %hu is too high, capping to MAX_ITEM_RDM_OPT...\n", amount );
```

File: src/map/itemdb.cpp [hunk 15]
```diff
@@ -1259,6 +1259,392 @@ int16 ItemGroupDatabase::item_exists_pc(map_session_data *sd, uint16 group_id)
...
+
+	if( this->nodeExists( node, "ResultRefine" ) ){
+		uint16 refine;
+
+		if( !this->asUInt16( node, "ResultRefine", refine ) ){
+			return 0;
+		}
+
+		if( refine > MAX_REFINE ){
+			this->invalidWarning( node["ResultRefine"], "laphine_upgrade_parseBodyNode: result refine %hu is too high, capping to MAX_REFINE...\n", refine );
```
Comment (aleos89) [inline src/map/itemdb.cpp:1590]:
```suggestion
			this->invalidWarning( node["ResultRefine"], "laphine_upgrade_parseBodyNode: Result refine %hu is too high, capping to MAX_REFINE...\n", refine );
```

File: src/map/itemdb.cpp [hunk 16]
```diff
@@ -1259,6 +1259,392 @@ int16 ItemGroupDatabase::item_exists_pc(map_session_data *sd, uint16 group_id)
...
+
+	if( this->nodeExists( node, "ResultRefineMinimum" ) ){
+		uint16 refine;
+
+		if( !this->asUInt16( node, "ResultRefineMinimum", refine ) ){
+			return 0;
+		}
+
+		if( refine > MAX_REFINE ){
+			this->invalidWarning( node["ResultRefineMinimum"], "laphine_upgrade_parseBodyNode: result refine minimum %hu is too high, capping to MAX_REFINE...\n", refine );
```
Comment (aleos89) [inline src/map/itemdb.cpp:1609]:
```suggestion
			this->invalidWarning( node["ResultRefineMinimum"], "laphine_upgrade_parseBodyNode: Result refine minimum %hu is too high, capping to MAX_REFINE...\n", refine );
```

File: src/map/itemdb.cpp [hunk 17]
```diff
@@ -1259,6 +1259,392 @@ int16 ItemGroupDatabase::item_exists_pc(map_session_data *sd, uint16 group_id)
...
+
+	if( this->nodeExists( node, "ResultRefineMaximum" ) ){
+		uint16 refine;
+
+		if( !this->asUInt16( node, "ResultRefineMaximum", refine ) ){
+			return 0;
+		}
+
+		if( refine > MAX_REFINE ){
+			this->invalidWarning( node["ResultRefineMaximum"], "laphine_upgrade_parseBodyNode: result refine maximum %hu is too high, capping to MAX_REFINE...\n", refine );
```
Comment (aleos89) [inline src/map/itemdb.cpp:1628]:
```suggestion
			this->invalidWarning( node["ResultRefineMaximum"], "laphine_upgrade_parseBodyNode: Result refine maximum %hu is too high, capping to MAX_REFINE...\n", refine );
```

**Review by aleos89 — APPROVED** [review]
Overall, looks great! Just some minor comments and thoughts.

**Comment by kennykiro1** [comment]
I'm getting this error in several items, how can I fix it?

**Comment by Lemongrass3110** [comment]
> 

@kennykiro1 works just fine for me. Check if you merged correctly or check if you have errors during loading in your mapserver.
grafik

grafik

**Comment by kennykiro1** [comment]
> > 
> 
> @kennykiro1 funciona muito bem para mim. Verifique se você fez a fusão corretamente ou verifique se há erros durante o carregamento no seu mapserver. gráfico
> 
> gráfico

is it because I didn't convert the item_combo_db.txt to yml?

**Comment by kennykiro1** [comment]
> 


- Item: Shadow_Refine_Hammer 
this and several others I can't upgrade :/

File: src/map/itemdb.cpp (hunk 6 above)
Comment (Lemongrass3110) [reply src/map/itemdb.cpp:1279]: Can be removed. I think I just copied and adjusted it when I originally wrote it.

File: db/laphine_synthesis.yml [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+#
+###########################################################################
+# Laphine Synthesis Database
+###########################################################################
+#
+# Laphine Synthesis Settings
+#
+###########################################################################
+# - Item                       Item that triggers Laphine Synthesis.
+#   RewardGroup                Item Group that will given out as a reward.
```
Comment (aleos89) [inline db/laphine_synthesis.yml:26]:
```suggestion
#   RewardGroup                Item Group that will be given out as a reward.
```

File: db/laphine_synthesis.yml [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+#   RequiredRequirementsCount  How many requirements have to be fulfilled? (Default: 1)
```
Comment (aleos89) [inline db/laphine_synthesis.yml:27]:
```suggestion
#   RequiredRequirementsCount  Amount of requirements that have to be fulfilled. (Default: 1)
```

File: db/laphine_synthesis.yml [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+#   MinimumRefine              Minimum refine level of the required items. (Default: 0)
+#   MaximumRefine              Maximum refine level of the required items. (Default: MAX_REFINE)
+#   Requirements:              List of possible item(s) as requirements
```
Comment (aleos89) [inline db/laphine_synthesis.yml:30]:
```suggestion
#   Requirements:              List of possible requirement items.
```

File: db/re/laphine_synthesis.yml [hunk 21]
```diff
@@ -0,0 +1,4492 @@
...
+#
+###########################################################################
+# Laphine Synthesis Database
+###########################################################################
+#
+# Laphine Synthesis Settings
+#
+###########################################################################
+# - Item                       Item that triggers Laphine Synthesis.
+#   RewardGroup                Item Group that will given out as a reward.
```
Comment (aleos89) [inline db/re/laphine_synthesis.yml]:
```suggestion
#   RewardGroup                Item Group that will be given out as a reward.
```

File: db/re/laphine_synthesis.yml [hunk 22]
```diff
@@ -0,0 +1,4492 @@
...
+#   RequiredRequirementsCount  How many requirements have to be fulfilled? (Default: 1)
```
Comment (aleos89) [inline db/re/laphine_synthesis.yml]:
```suggestion
#   RequiredRequirementsCount  Amount of requirements that have to be fulfilled. (Default: 1)
```

File: db/re/laphine_synthesis.yml [hunk 23]
```diff
@@ -0,0 +1,4492 @@
...
+#   MinimumRefine              Minimum refine level of the required items. (Default: 0)
+#   MaximumRefine              Maximum refine level of the required items. (Default: MAX_REFINE)
+#   Requirements:              List of possible item(s) as requirements
```
Comment (aleos89) [inline db/re/laphine_synthesis.yml]:
```suggestion
#   Requirements:              List of possible requirement items.
```

File: db/re/laphine_synthesis.yml [hunk 24]
```diff
@@ -0,0 +1,4492 @@
...
+#     - Item                   Item name of the required item
+#       Amount                 How many of the specific item are required? (Default: 1)
```
Comment (aleos89) [inline db/re/laphine_synthesis.yml]:
```suggestion
#       Amount                 Amount of specific required item. (Default: 1)
```

File: db/laphine_synthesis.yml [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+#     - Item                   Item name of the required item
+#       Amount                 How many of the specific item are required? (Default: 1)
```
Comment (aleos89) [inline db/laphine_synthesis.yml:32]:
```suggestion
#       Amount                 Amount of specific required item. (Default: 1)
```

**Review by aleos89 — APPROVED** [review]
Couple more comments to database header fixes.

**Pull request merged**